Redirect upon mod creation and upload

TheUnifox opened the pull request:
Redirect the user upon creating a mod, and upon uploading a mod version. This follows what the confirmation popups say should happen upon hitting ok. Also increase mod file upload limit from 75mb to 150mb. I'm gonna need a larger upload limit eventually for libcef being over 100mb

TheUnifox committed:
Redirect upon submission

Also give a bit more info about mod description when uploading

TheUnifox committed:
Increase files size pls

75mb is enough for most things, but 150 would allow some larger things like libcef I will need to upload eventually

Futuremappermydud commented:
![image](https://github.com/user-attachments/assets/22d6903a-f69b-442d-84a1-e206df1cfcc6)
The CEF binary is less than 75mb so I don't see why this is an issue. The current limit is already generous considering most mods barely even hit 1mb.

Caeden117 commented on src/routes/upload/version/[id]/+page.svelte:
```diff
@@ -353,7 +354,7 @@
           const target = e.target as HTMLInputElement;
           if (target && target.files && target.files.length == 1) {
-            if (target.files[0].size > 75 * 1024 * 1024) {
+            if (target.files[0].size > 150 * 1024 * 1024) {
```
Would it be a good idea to consider pulling this hardcoded number into a configuration setting of some kind? I imagine there will be some discussion regarding the maximum file size for an uploaded mod, and it'll definitely be easier to change the config file than changing this magic number every time.

Futuremappermydud replied:
Probably but the frontend doesn't have much configuration other than api url

TheUnifox commented:
> ![image](https://private-user-images.githubusercontent.com/54294576/408808601-22d6903a-f69b-442d-84a1-e206df1cfcc6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg0NDczODgsIm5iZiI6MTczODQ0NzA4OCwicGF0aCI6Ii81NDI5NDU3Ni80MDg4MDg2MDEtMjJkNjkwM2EtZjY5Yi00NDJkLTg0YTEtZTIwNmRmMWNmY2M2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjAxVDIxNTgwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYyY2JmMmQ0YzhkMjMwNTRjOWY4OWJlOGQ3MmNkZTlmMWEzNTY3NzAzNDdkNGIyNTM3NzlkNTllODEyMmIzYjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.W3ROSypu1ZzI8N3-OjMtRVr5mgksDKViwRxjvEcNB4k) The CEF binary is less than 75mb so I don't see why this is an issue. The current limit is already generous considering most mods barely even hit 1mb.

That appears to be a libcef_*wrapper* you found. libcef is about 209mb + the data files it requires. Which does compress down to a 107mb zip. Built libcef versions are available [here](https://cef-builds.spotifycdn.com/index.html), of which the version I require is 120.1.8+ge6b45b0+chromium-120.0.6099.109 / Chromium 120.0.6099.109 (Standard Distrib) You have to hit show all builds, show more builds, and scroll down quite a ways to find it.
![image](https://github.com/user-attachments/assets/3c6e19dc-8a3d-44a7-a7be-2b84b4dba56f)
![image](https://github.com/user-attachments/assets/f266e59a-e0a9-4dfd-b9bb-89a30f351b5a)

Though the upload limit isn't a huge issue yet, and probably wont be for a while. Cef is really difficult to work with, and I'm stuck on a strange issue that I have no idea why it's happening. So it'll be a while until I need to upload it. 
Also I know relative to most other mods and libs it's massive, but that's just cef for you. I've pulled the cef source to hunt through and maybe even rewrite some of to integrate directly with Unity, but it's literally over 1.2 million files, a total of over 180gb. Luckily I don't think I have to go through ***everything***, but some major parts probably have to be looked through at least. 
I at least have to look through one specific part to see why it's doing something. Then *maybe* in the future looking through some of the rendering code to see if I can get it to use Unity to render, instead of the multiple renderer libraries it depends on.
Hell, maybe I could contact the lead dev and project manager of cef himself to ask what's going on... idk

But anyways, that'll take a while, so I won't need the increased limit for quite a while. But pulling it out into a config file is a good idea, Hardcoded magic numbers like this usually aren't great to use.

TheUnifox committed:
Move upload limit to .env

Futuremappermydud commented:
Ah, okay. I probably won't merge this now without further discussion. Large files like this aren't good for users either, at least not without a warning, so we will probably need more backend changes to add some size marker or something.

TheUnifox commented:
now the upload limit is set by PUBLIC_FILE_UPLOAD_LIMIT_MB env. this also requires [pull #43](https://github.com/Saeraphinx/BadBeatMods/pull/43) in the backend to work.

Probably would be good to have the file size shown, and maybe a warning if it's over a certain size, like 50mb or sumn

TheUnifox committed:
add file size and oversize warning

TheUnifox commented:
using the recent change to the backend pr, show mod file size, and a warning if over a limit

TheUnifox committed:
Check for LargeFiles role to allow oversized files

TheUnifox commented:
Keep the limit at current 75mb. Now just add the LargeFiles role to a user to allow them to exceed the limit when needed. Currently allows up to 3x the limit with the backend changes

TheUnifox committed:
Move file size + change icon

Looks much better put below the name and summary then over at the right icons. Also catch cases where a mod doesn't have a fileSize, showing 0KB instead of NaNKB. Also also, showing the oversize warning next to the size, instead of next to the name

TheUnifox committed:
Merge branch 'main' into main

TheUnifox committed:
Add file size + Adjust oversize warning icon

Show a file size for each modVersion, instead of just the mod as a whole. And adjust the look of the oversize warning a little.
Also a bit of the start of next commit... oops

TheUnifox committed:
Mods won't have fileSize, only modVersions

Because of some changes to the backend PRs, (the PR was split out into multiple), mods won't have a fileSize. So fileSize has to be taken from a modVersion.
This is better anyways, that way instead of always being the latest modVersion, it takes the latest modVersion available for a version. This makes the fileSize much more accurate if the mods size changes a lot between versions

TheUnifox commented:
Moved the file size and oversize warning around a bit, and adjusted for the backend PRs changes. The backend PR has been split out into multiple PRs (I have yet to PR the branches tho). Once each of them are pulled, this PR will be fully functional. It *is* completely functional, and doesn't give any errors, But the file sizes just show 0KB until the backend is changed.
So this can be accepted without breaking anything, just file sizes wont work *quite* yet

TheUnifox commented:
Also, file sizes might take a bit on the backend, since it makes database changes. That needs migration code before it can be added

TheUnifox committed:
minor fixes

Just some minor fixes to get it fully working. Now tested with an updated backend to make sure file sizes work!

TheUnifox commented:
Made some minor fixes. Database migration for the backend is done and working, and I've tested that file sizes will fully work once the backend is updated

TheUnifox committed:
Merge branch 'main' into main

TheUnifox commented:
Resolve conflicts caused by recent commits. Needs testing and further changes to ensure nothing has broken

TheUnifox committed:
Fix file size on mod page caused by recent commits

TheUnifox committed:
Add size limit to .env

TheUnifox commented:
Fix a small issue caused by recent commits, and make sure the .env is correct

TheUnifox committed:
Merge branch 'Futuremappermydud:main' into main

TheUnifox committed:
Merge branch 'Futuremappermydud:main' into main

TheUnifox commented:
Just keeping up to date until backend updates

TheUnifox commented:
Last PR on the backend adding file sizes has been pulled. Once the backend updates I'll do some testing to make sure it's all working properly

TheUnifox commented:
Backend has updated, and everything is working perfectly. If there aren't anymore changes needed I think this is ready to be pulled.

Saeraphinx commented on example.env:
```diff
@@ -1,2 +1,3 @@
-PUBLIC_API_BASE_URL="http://localhost:5001/"
\ No newline at end of file
+PUBLIC_API_BASE_URL="http://localhost:5001/"
+FILE_UPLOAD_LIMIT_MB="150"
```
default limit is now 50MB (large files role has a default multiplier of 3x, meaning those with the role have a maximum of 50mb)

Saeraphinx commented on src/lib/components/ui/mods/ModCardBase.svelte:
```diff
@@ -72,6 +75,23 @@
       <p class="text-left text-xs text-neutral-foreground-3 md:text-sm">
         {mod.summary}
       </p>
+      <div class="flex w-fit flex-row gap-2">
+        <svg viewBox="0 0 20 20" class="w-4 text-neutral-foreground-2 md:w-4">
+          <DatabaseArrowDownFilled/>
+        </svg>
+        <span
+          class="silly-capitalize text-sm text-neutral-foreground-2 md:text-base"
+          >{latestSize ? (latestSize /1024 > 1024 ? `${Math.round(latestSize/1024/1024)}MB` : (`${Math.round(latestSize/1024)}KB` === `NaNKB` ? `0KB` : `${Math.round(latestSize/1024)}KB`)) : `0KB`}</span
+        >
+        {#if latestSize && latestSize/1024/1024 > 50}
+          <span class="text-lg mt-auto h-fit w-fit tooltip">
+            <svg viewBox="0 0 30 30" class="w-6 text-neutral-foreground-2 text-yellow-400 md:w-6">
+              <WarningFilled />
+            </svg>
+            <span class="tooltiptext">This is a very large mod!</span>
+          </span>
+        {/if}
+      </div>
```
Mods do not have a fileSize, it does not make sense to me to have the base card display a size

TheUnifox replied:
It's put there so the latest mod version size shows in the mod list, and for approvers to see a version's size. It's too much over on the right for the size in the mods list, and would also be too crowded to put it in with the other info on the right on the approval card. It may not really be *needed* in the mod list, it's usually just other mod devs going to the site anyways. But I think it would be good to have it for approvers to see how large a mod is. And unless having a whole separate modcardbase just for version approval would be fine, having filesize here is the only way it looks decent to have the size for approvers. I could make another ModCardBase with fileSize just for version approval to use, probably called ModVersionApprovalCardBase. Then fileSize can be removed from ModCardBase, and still have fileSize for approvers. Approvers having fileSize could be useful for a couple reasons. Seeing a very large fileSize could let them know it might be better to just check hashes, and not bother looking through a possibly *massive* source repository. It could also give them a chance to contact the mod dev to ask why it's so massive, instead of just approving it with no idea how big it is. If they contact the dev, and there isn't really a reason for it to be so massive, then the dev can slim it down before uploading it again, keeping from approving an unnecessarily large mod.

If everyone agrees that having fileSize for the approvers isn't really needed, I'll remove it from here. else it'll stay. Having it for the approvers is the main reason to have it here, and having it here also happens to show it in the mod list, which can be nice for anyone looking through it, including mod devs to see if maybe a lib they were going to use is actually rather large. It does also add it to the top of the mod page and a couple other spots which really isn't needed at all, but it's not harming anything being there, just a little bit of extra info ig.

~~sorry, after midnight rant time :3~~

TheUnifox replied:
also btw, noticed you added an endpoint to the backend to update all fileSizes that are 0, is that going to be used soon for accurate fileSizes to be shown when this and the bsmanager pr are pulled? Not sure exactly what the bsmanager pr is waiting on, but it might be fore fileSizes to actually show accurately, and not just 0

Saeraphinx replied:
> And unless having a whole separate modcardbase just for version approval would be fine, having filesize here is the only way it looks decent to have the size for approvers. I could make another ModCardBase with fileSize just for version approval to use, probably called ModVersionApprovalCardBase. Then fileSize can be removed from ModCardBase, and still have fileSize for approvers.

Would having the file size hide itself if it is null/undefined/0/a falsy value (as a way to hide it) be a good compromise? I agree that it would be a thing approvers might want to be able to see, but I don't think its something that should be visible without a download button next to it...

TheUnifox replied:
ahhh yes, that would be good

TheUnifox replied:
Ya that's much better

Saeraphinx replied:
> also btw, noticed you added an endpoint to the backend to update all fileSizes that are 0, is that going to be used soon for accurate fileSizes to be shown when this and the bsmanager pr are pulled? Not sure exactly what the bsmanager pr is waiting on, but it might be fore fileSizes to actually show accurately, and not just 0

eventually, I wanted to test the new admin panel with that.

Saeraphinx commented on src/lib/components/ui/versions/VersionCard.svelte:
```diff
@@ -95,6 +95,18 @@
+        {(version.fileSize /1024 > 1024 ? `${Math.round(version.fileSize/1024/1024)}MB` : (`${Math.round(version.fileSize/1024)}KB` === `NaNKB` ? `0KB` : `${Math.round(version.fileSize/1024)}KB`))}
+      </div>
+
+      {#if version.fileSize/1024/1024 > 50}
```
Might make sense to make this dynamic with the env variable for the regular limit? or maybe seperate it out into a different env variable.

Saeraphinx commented on src/lib/types/Approval.ts:
```diff
@@ -19,6 +19,7 @@ export interface ModVersionDBObject {
   downloadCount: number;
   lastApprovedById: number | null;
   lastUpdatedById: number;
+  fileSize: number;
```
`fileSize` needs to be omitted by accompanying approval object on L46

Saeraphinx reviewed (changes requested):
I'd like to see this get merged ASAP once the things mentioned above are resolved, as those redirects that are also contained in this PR are heavily needed.

Saeraphinx committed:
Merge remote-tracking branch 'origin/main' into pr/TheUnifox/10-1

TheUnifox committed:
File upload size limit now 50MB

TheUnifox committed:
Use upload limit as warning size

TheUnifox committed:
omit fileSize

no need for it here, keeps it from being edited falsely

TheUnifox committed:
Dont show fileSize where not needed

Saeraphinx committed:
make property optional

Saeraphinx committed:
remove redundant filesize marker
